[SPARK-13222][Streaming][WIP]make sure latest status of stateful RDD be checkpointed on gracefulshutdown

mwws opened the pull request:
The CheckpointInterval of DStream could be customized to multiple of BatchInteveal. So when user call shutdown it could be in the middle of two RDD checkpoint. In case the input source is not repeatable and user don't want to enable WAL because of its extra cost, the application could be unrecoverable.

This PR is to make sure the latest status of stateful RDD could be checkpointed on graceful shutdown

wei-mao-intel committed:
make sure latest status of stateful RDD be checkpointed on gracefulshutdown

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/DStreamGraph.scala:
```diff
@@ -100,6 +100,13 @@ final private[streaming] class DStreamGraph extends Serializable with Logging {
   def getOutputStreams(): Array[DStream[_]] = this.synchronized { outputStreams.toArray }
 
+  def readyToShutdown() : Unit = this.synchronized {
+    this.synchronized {
```
typo? remove the extra `this.synchronized {`

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/DStreamGraph.scala:
```diff
@@ -131,6 +138,10 @@ final private[streaming] class DStreamGraph extends Serializable with Logging {
     logDebug("Cleared old metadata for time " + time)
   }
 
+  def isCheckpointMissedLastTime() : Boolean = {
```
Nit: remove the space before `:`

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/dstream/DStream.scala:
```diff
@@ -490,6 +494,19 @@ abstract class DStream[T: ClassTag] (
+  private[streaming] def readyToShutdown(): Unit = {
+    _readyToShutdown = true
+    dependencies.foreach(_.readyToShutdown())
+    logDebug("Ready to shutdown")
```
Will it cause many log output since it's called in a recursive way? Since we have one in `DStreamGraph`, can we remove this?

mwws replied:
It's in debug level, and you can refer to other method in this file, they all have the same behavior. I think it should be fine.

chenghao-intel replied:
yes I know, I mean you have one https://github.com/apache/spark/pull/11101/files#diff-221bc6301915f7a476786c794b855b21R105 already which print out exactly the same log, probably we can add more information in this log? for example: creationSite?

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/DStreamGraph.scala:
```diff
@@ -131,6 +138,10 @@ final private[streaming] class DStreamGraph extends Serializable with Logging {
   }
 
+  def isCheckpointMissedLastTime() : Boolean = {
+    outputStreams.foldLeft(false)((value, next) => value || next.isCheckpointMissedLastTime)
```
call the `outputStreams.isCheckpointMissedLastTime()` directly?

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -123,6 +126,12 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
 
+      // generate one more bacth to make sure RDD in lastJob is checkpointed.
+      if (!jobScheduler.receiverTracker.hasUnallocatedBlocks &&
+        ssc.graph.isCheckpointMissedLastTime) {
```
Can you add more comment to explain why we need to check the `ssc.graph.isCheckpointMissedLastTime`?

wei-mao-intel committed:
fix typo and remove extra space

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50862/
Test FAILed.

SparkQA commented:
**[Test build #50863 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50863/consoleFull)** for PR 11101 at commit [`4f2efe4`](https://github.com/apache/spark/commit/4f2efe4962c01a83de600d2651463c1fb63038cb).

SparkQA commented:
**[Test build #50863 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50863/consoleFull)** for PR 11101 at commit [`4f2efe4`](https://github.com/apache/spark/commit/4f2efe4962c01a83de600d2651463c1fb63038cb).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50863/
Test FAILed.

zsxwing commented:
I'm not sure if this is really helpful. In general, if the user doesn't enable WAL and also doesn't use a replayable source, I don't think the user will care about losing data.

mwws commented:
@zsxwing, we have real use case that data integrity is important, but at the same time, we can not afford extra time cost of WAL with limited data processing power/host. (Batch processing time is already very closed to BatchInterval at peak time.) 

And anyway, this PR is making the best effort to avoid data losing with little overhead. I think it's worth to do.

chenghao-intel commented:
Some of users claims that they will use Kafka, but without the WAL for performance concern, see comments at https://github.com/apache/spark/pull/10252#issuecomment-184515856

People have workaround solution, but it's better if Spark Streaming can support that internally.

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -123,6 +126,12 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
+      // generate one more bacth to make sure RDD in lastJob is checkpointed.
+      if (!jobScheduler.receiverTracker.hasUnallocatedBlocks &&
+        ssc.graph.isCheckpointMissedLastTime) {
+        Thread.sleep(ssc.graph.batchDuration.milliseconds)
```
What about if the last batch execution takes longer time than the batchDuration? Probably we need a notification with a timeout?

mwws replied:
If task execution time is longer than batchDuration, it usually means that the streaming application is not healhy and user need to adjust parameters to avoid such scenario.

And the "thread.sleep" here is to make sure a new job could be generated, it's nothing to do with job execution.

chenghao-intel replied:
ok, I was wondering it will be more graceful if this thread can be woke up as soon as the batch job finished, other than sleep with a fixed length of time.

chenghao-intel commented:
Sorry, I mean we'd better try our best to save the stateful data as much as we can, no matters whether if it's the replayable data source or WAL enabled etc.

wei-mao-intel committed:
fix bug and enhance log

1. fix a bug about calling max on empty seq
2. add meaningful information into log
3. enhance comment

SparkQA commented:
**[Test build #51947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51947/consoleFull)** for PR 11101 at commit [`cef511c`](https://github.com/apache/spark/commit/cef511c497d664ac33a87b5f1b40d07dfe695cdf).

SparkQA commented:
**[Test build #51947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51947/consoleFull)** for PR 11101 at commit [`cef511c`](https://github.com/apache/spark/commit/cef511c497d664ac33a87b5f1b40d07dfe695cdf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51947/
Test PASSed.

chenghao-intel commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala:
```diff
@@ -123,6 +126,15 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
+      if (!jobScheduler.receiverTracker.hasUnallocatedBlocks &&
+        ssc.graph.isCheckpointMissedLastTime) {
+        Thread.sleep(ssc.graph.batchDuration.milliseconds)
+      }
```
I just wonder if we can add a double check if the last mini batch finish, otherwise, at least we can add a warning log.
